doc/manual/interfaces.rst

```diff
@@ -0,0 +1,278 @@
+
+Returning to our collection of squares from above, we could instead define it as a subtype of an ``AbstractArray``:
+
+.. doctest::
+
+    julia> immutable SquaresVector <: AbstractArray{Int, 1}
+               count::Int
+           end
+           Base.size(S::SquaresVector) = (S.count,)
+           Base.linearindexing(::Type{SquaresVector}) = Base.LinearFast()
+           Base.getindex(S::SquaresVector, i::Int) = i*i;
+
+Note that it's very important to specify the two parameters of the ``AbstractArray``; the first defines the :func:`eltype`, and the second defines the :func:`ndims`.  But that's it takes for our squares type to be an iterable, indexable, and completely functional array:
+
+.. testsetup::
+
+    srand(1);
+
+.. doctest::
+
+    julia> s = SquaresVector(7)
+    7-element SquaresVector:
+      1
+      4
+      9
+     16
+     25
+     36
+     49
+
+    julia> s[s .> 20]
+    3-element Array{Int64,1}:
+     25
+     36
+     49
+
+    julia> s \ rand(7,2)
+    1x2 Array{Float64,2}:
+     0.0151876  0.0179393
+
+As a more complicated example, let's define our own toy N-dimensional sparse-like array type built on top of ``Dict``:
+
+.. doctest::
+
+    julia> immutable SparseArray{T,N} <: AbstractArray{T,N}
+               data::Dict{NTuple{N,Int}, T}
+               dims::NTuple{N,Int}
+           end
+           SparseArray{T}(::Type{T}, dims::Int...) = SparseArray(T, dims)
+           SparseArray{T,N}(::Type{T}, dims::NTuple{N,Int}) = SparseArray{T,N}(Dict{NTuple{N,Int}, T}(), dims)
+    SparseArray{T,N}
+
+    julia> Base.size(A::SparseArray) = A.dims
+           Base.similar{T}(A::SparseArray, ::Type{T}, dims::Dims) = SparseArray(T, dims)
+           # Define scalar indexing and indexed assignment up to 3-dimensions
+           Base.getindex{T}(A::SparseArray{T,1}, i1::Int)                   = get(A.data, (i1,), zero(T))
+           Base.getindex{T}(A::SparseArray{T,2}, i1::Int, i2::Int)          = get(A.data, (i1,i2), zero(T))
+           Base.getindex{T}(A::SparseArray{T,3}, i1::Int, i2::Int, i3::Int) =  get(A.data, (i1,i2,i3), zero(T))
+           Base.setindex!{T}(A::SparseArray{T,1}, v, i1::Int)                   = (A.data[(i1,)] = v)
+           Base.setindex!{T}(A::SparseArray{T,2}, v, i1::Int, i2::Int)          = (A.data[(i1,i2)] = v)
+           Base.setindex!{T}(A::SparseArray{T,3}, v, i1::Int, i2::Int, i3::Int) = (A.data[(i1,i2,i3)] = v);
+
+Notice that this is a ``LinearSlow`` array, so we must manually define :func:`getindex` and :func:`setindex!` for each dimensionality we'd like to support.  Unlike the ``SquaresVector``, we are able to define :func:`setindex!`, and so we can mutate the array:
```

Comment:
Maybe you should way a word about this difference between ``LinearSlow`` and ``LinearFast`` right after the table above?